sentry/src/routes/campaign.rs

```diff
@@ -25,32 +68,197 @@ pub async fn create_campaign<A: Adapter>(
-    //             PoolError::Backend(error) if error.code() == Some(&SqlState::UNIQUE_VIOLATION) => {
-    //                 Err(ResponseError::Conflict(
-    //                     "channel already exists".to_string(),
-    //                 ))
-    //             }
-    //             _ => Err(error_response),
-    //         }
-    //     }
-    //     Ok(false) => Err(error_response),
-    //     _ => Ok(()),
-    // }?;
-
-    let create_response = SuccessResponse { success: true };
+    if campaign.budget > remaining_for_channel {
+        return Err(ResponseError::BadRequest("Not Enough budget for campaign".to_string()));
+    }
+
+    // If the channel is being created, the amount spent is 0, therefore remaining = budget
+    increase_remaining_for_campaign(&app.redis.clone(), campaign.id, campaign.budget).await.map_err(|_| ResponseError::BadRequest("Couldn't update remaining while creating campaign".to_string()))?;
+
+    // insert Campaign
+    match insert_campaign(&app.pool, &campaign).await {
+        Err(error) => {
+            error!(&app.logger, "{}", &error; "module" => "create_campaign");
+            match error {
+                PoolError::Backend(error) if error.code() == Some(&SqlState::UNIQUE_VIOLATION) => {
+                    Err(ResponseError::Conflict(
+                        "Campaign already exists".to_string(),
+                    ))
+                }
+                _ => Err(error_response),
+            }
+        }
+        Ok(false) => Err(ResponseError::BadRequest("Encountered error while creating Campaign; please try again".to_string())),
+        _ => Ok(()),
+    }?;
 
     Ok(success_response(serde_json::to_string(&campaign)?))
 }
 
+
+// tested
+fn get_total_remaining_for_channel(accounting_spent: &UnifiedNum, latest_spendable: &Spendable) -> Option<UnifiedNum> {
+    let total_deposited = latest_spendable.deposit.total;
+
+    let total_remaining = total_deposited.checked_sub(&accounting_spent);
+    total_remaining
+}
+
+pub mod update_campaign {
+    use super::*;
+    use lazy_static::lazy_static;
+
+    lazy_static!{
+        pub static ref CAMPAIGN_REMAINING_KEY: &'static str = "campaignRemaining";
+    }
+
+    pub async fn increase_remaining_for_campaign(redis: &MultiplexedConnection, id: CampaignId, amount: UnifiedNum) -> Result<bool, CampaignError> {
+        let key = format!("{}:{}", *CAMPAIGN_REMAINING_KEY, id);
+        redis::cmd("INCRBY")
+            .arg(&key)
+            .arg(amount.to_u64())
+            .query_async(&mut redis.clone())
+            .await?;
+        Ok(true)
+    }
+
+    pub async fn decrease_remaining_for_campaign(redis: &MultiplexedConnection, id: CampaignId, amount: UnifiedNum) -> Result<bool, CampaignError> {
+        let key = format!("{}:{}", *CAMPAIGN_REMAINING_KEY, id);
+        redis::cmd("DECRBY")
+            .arg(&key)
+            .arg(amount.to_u64())
+            .query_async(&mut redis.clone())
+            .await?;
+        Ok(true)
```

Comment:
The redis command will return the new value. We must check if this value is still positive when modifying the Campaign Budget!

- [ ] Return the value from the redis commmand
- [ ] Write a validation that it is still `> 0` when modifying campaign